src/hotspot/share/nmt/virtualMemoryTracker.cpp

```diff
@@ -485,19 +485,25 @@ bool VirtualMemoryTracker::remove_released_region(ReservedMemoryRegion* rgn) {
   return result;
 }
 
-bool VirtualMemoryTracker::remove_released_region(address addr, size_t size) {
+bool VirtualMemoryTracker::remove_released_region(address addr, size_t size, bool extra_memory) {
   assert(addr != nullptr, "Invalid address");
   assert(size > 0, "Invalid size");
   assert(_reserved_regions != nullptr, "Sanity check");
 
   ReservedMemoryRegion  rgn(addr, size);
   ReservedMemoryRegion* reserved_rgn = _reserved_regions->find(rgn);
 
-  if (reserved_rgn == nullptr) {
+  if (reserved_rgn == nullptr && extra_memory) {
     log_debug(nmt)("No reserved region found for (" INTPTR_FORMAT ", " SIZE_FORMAT ")!",
                   p2i(rgn.base()), rgn.size());
+    return true;
   }
   assert(reserved_rgn != nullptr, "No reserved region");
+  if (extra_memory) {
+    if (reserved_rgn->flag() != mtClassShared)
+      return true;
+    assert(reserved_rgn->end() == rgn.end() || reserved_rgn->base() == rgn.base(), "extra memory should be at either end of the region.");
+  }
```

Comment:
This seems like an extreme hack. I understand that this just follows the tradition of the rest of the hacks in this file, but can't this be better handled in the CDS layer above?

Reply:
I'd like to take an extra look at that before this PR gets integrated.

Reply:
I agree and will work on it.